### PR TITLE
support custom hash partition

### PR DESCRIPTION
**Pull request by timchenxiaoyu**
allow user  custom define hash key

### CODE REVIEW

File: plugins/flusher/kafka/flusher_kafka.go [hunk 1]
```diff
@@ -17,15 +17,20 @@ package kafka
 import (
 	"encoding/json"
 	"errors"
-	"time"
-
 	"github.com/alibaba/ilogtail"
 	"github.com/alibaba/ilogtail/pkg/logger"
 	"github.com/alibaba/ilogtail/pkg/protocol"
+	"strings"
```
Comment (shabicheng) [inline plugins/flusher/kafka/flusher_kafka.go:23]: make lint

File: plugins/flusher/kafka/flusher_kafka.go [hunk 2]
```diff
@@ -104,17 +112,37 @@ func (k *FlusherKafka) Description() string {
...
+					for _, content := range log.GetContents() {
+						if contains(k.HashKey, content.Key) {
+							hashData = append(hashData, content.Value)
+						}
+					}
+					if len(hashData) == 0 {
+						hashData = append(hashData, logstoreName)
+					}
+					logger.Debug(k.context.GetRuntimeContext(), "partition key", hashData)
+					hashKey = sarama.StringEncoder(strings.Join(hashData, ""))
```
Comment (shabicheng) [inline plugins/flusher/kafka/flusher_kafka.go:136]: strings.Join(hashData, "###")

File: plugins/flusher/kafka/flusher_kafka.go [hunk 3]
```diff
@@ -33,11 +33,17 @@ type FlusherKafka struct {
 	SASLPassword    string
 	Topic           string
 	PartitionerType string
+	HashKeys        []string
+	HashOnce        bool
 	ClientID        string
```
Comment (shabicheng) [inline plugins/flusher/kafka/flusher_kafka.go:38]: add a new empty line before isTerminal bool

File: plugins/flusher/kafka/flusher_kafka.go (hunk 1 above)
Comment (timchenxiaoyu) [reply plugins/flusher/kafka/flusher_kafka.go:20]: change gofmt to goimport

File: plugins/flusher/kafka/flusher_kafka.go [hunk 4]
```diff
@@ -102,18 +113,51 @@ func (k *FlusherKafka) Description() string {
 }
 
 func (k *FlusherKafka) Flush(projectName string, logstoreName string, configName string, logGroupList []*protocol.LogGroup) error {
+	var flusher FlusherFunc
```
Comment (shabicheng) [inline plugins/flusher/kafka/flusher_kafka.go:116]: move these codes to Init

```
	if k.PartitionerType == "hash" {
		k.flusher = k.HashFlush
	} else {
		k.flusher = k.NormalFlush
	}
```

**Pull request merged**